### PR TITLE
SymmetricTensor and Tensor class AD number compatibility

### PR DESCRIPTION
**Pull request by jppelteret**
This sequence of commits generalises the `SymmetricTensor` class a little more, as well as providing it and the `Tensor` class the modifications necessary to support additional AD number types. This mainly centres around the safe initialisation of all underlying values, but also includes changes that minimise the number of created temporary variables.

### CODE REVIEW

**Comment by jppelteret** [comment]
In reference to #4704

File: include/deal.II/base/symmetric_tensor.h [hunk 1]
```diff
@@ -539,7 +539,7 @@ class SymmetricTensor
   /**
    * Default constructor. Creates a tensor with all entries equal to zero.
    */
-  SymmetricTensor () = default;
```
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: ~~why did you remove `= default` keyword?~~ ah, i see below that you need explicitly initialize to zero AD numbers.

File: include/deal.II/base/symmetric_tensor.h [hunk 2]
```diff
@@ -580,12 +580,20 @@ class SymmetricTensor
...
+  template <typename OtherNumber>
+  SymmetricTensor &operator = (const SymmetricTensor<rank,dim,OtherNumber> &rhs);
+
+  /**
    * This operator assigns a scalar to a tensor. To avoid confusion with what
    * exactly it means to assign a scalar value to a tensor, zero is the only
    * value allowed for <tt>d</tt>, allowing the intuitive notation
    * <tt>t=0</tt> to reset all elements of the tensor to zero.
    */
-  SymmetricTensor &operator = (const Number d);
```
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: would not that disallow `Tensor<1,dim> r = 0`?

File: include/deal.II/base/symmetric_tensor.h [hunk 3]
```diff
@@ -606,35 +614,27 @@ class SymmetricTensor
...
   /**
    * Scale the tensor by <tt>factor</tt>, i.e. multiply all components by
    * <tt>factor</tt>.
    */
-  SymmetricTensor &operator *= (const Number factor);
+  template <typename OtherNumber>
+  SymmetricTensor &operator *= (const OtherNumber &factor);
 
   /**
    * Scale the vector by <tt>1/factor</tt>.
```
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: while here, you can fix `vector` to be `tensor`

File: include/deal.II/base/symmetric_tensor.h [hunk 4]
```diff
@@ -606,35 +614,27 @@ class SymmetricTensor
...
-   */
-  SymmetricTensor   operator + (const SymmetricTensor &s) const;
-
-  /**
-   * Subtract two tensors. If possible, you should use <tt>operator -=</tt>
-   * instead since this does not need the creation of a temporary.
-   */
-  SymmetricTensor   operator - (const SymmetricTensor &s) const;
+  template <typename OtherNumber>
+  SymmetricTensor &operator /= (const OtherNumber &factor);
```
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: curious here as well about the change from by value to by reference

File: include/deal.II/base/symmetric_tensor.h [hunk 5]
```diff
@@ -606,35 +614,27 @@ class SymmetricTensor
...
   /**
    * Scale the vector by <tt>1/factor</tt>.
    */
-  SymmetricTensor &operator /= (const Number factor);
-
-  /**
-   * Add two tensors. If possible, you should use <tt>operator +=</tt> instead
-   * since this does not need the creation of a temporary.
-   */
-  SymmetricTensor   operator + (const SymmetricTensor &s) const;
```
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: looks like you removed `operator +` and `operator -` and introduce an operator for two tensors to add/subtract them.

Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: Thats right. This now mimics the `Tensor` class (see [here](https://github.com/dealii/dealii/blob/master/include/deal.II/base/tensor.h#L1355)), and is correct because the return type of the operation is not necessarily equal to that of the LHS operand.

File: include/deal.II/base/symmetric_tensor.h (hunk 1 above)
Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: Exactly

File: include/deal.II/base/symmetric_tensor.h (hunk 2 above)
Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: This is `SymmetricTensor`, so no ;-) But I don't think that changing this to a pass by reference is would stop this working.

File: include/deal.II/base/symmetric_tensor.h (hunk 4 above)
Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: The modification is in place simply to prevent the unnecessary creation of temporary variables. This is critical when working with AD numbers. But I am aware that the same might not be true for VectorizedArray. Do you see us doing this as a problem?

File: include/deal.II/base/symmetric_tensor.h (hunk 2 above)
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: > But I don't think that changing this to a pass by reference is would stop this working.

i guess you are right. I was just not sure about it myself.

File: include/deal.II/base/symmetric_tensor.h (hunk 4 above)
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: > Do you see us doing this as a problem?

i don't think so.

**Comment by davydden** [comment]
failing tests seems unrelated:
```
The following tests FAILED:
	 51 - numerics/create_laplace_matrix_constraints_01.debug (Failed)
	 88 - numerics/create_mass_matrix_constraints_01.debug (Failed)
	118 - numerics/create_laplace_matrix_constraints_02.debug (Failed)
	130 - numerics/create_mass_matrix_constraints_02b.debug (Failed)
	150 - numerics/create_mass_matrix_constraints_02.debug (Failed)
	162 - numerics/create_laplace_matrix_constraints_02b.debug (Failed)
	1849 - base/anisotropic_1.debug (Failed)
```

**Comment by jppelteret** [comment]
@Rombur @kronbichler @tamiko It would be good to know that all three of you are happy with the updates made here. It always feels like I'm walking into a mine-field when I fiddle with the `Tensor`/`SymmetricTensor` classes :-)

**Comment by jppelteret** [comment]
@davydden I agree - maybe these changes could have affected the result of `base/anisotropic_1.debug` (there's just a minor difference in the output)
```
----------------
##17      #:8   <== -52.8413
##17      #:8   ==> -52.8412
@ Absolute error = 1.0000000000e-4, Relative error = 1.8924626996e-6
----------------
##18      #:8   <== -52.8413
##18      #:8   ==> -52.8412
@ Absolute error = 1.0000000000e-4, Relative error = 1.8924626996e-6
----------------
##19      #:8   <== -3.89813
##19      #:8   ==> -3.89812
@ Absolute error = 1.0000000000e-5, Relative error = 2.5653391891e-6
##19      #:38  <== -1.00238
##19      #:38  ==> -1.00237
@ Absolute error = 1.0000000000e-5, Relative error = 9.9763560362e-6
##19      #:44  <== -1.00238
##19      #:44  ==> -1.00237
@ Absolute error = 1.0000000000e-5, Relative error = 9.9763560362e-6
##19      #:46  <== -1.00238
##19      #:46  ==> -1.00237
@ Absolute error = 1.0000000000e-5, Relative error = 9.9763560362e-6

+++  File "/deal.II/dealii/tests/base/anisotropic_1.output" differs from file "output"
```
but, at first glance, I don't see  how this touches the others.

File: include/deal.II/base/vectorization.h [hunk 6]
```diff
@@ -2707,6 +2707,25 @@ vectorized_transpose_and_store(const bool                    add_into,
...
 
 /**
+ * Relational operator == for VectorizedArray
+ *
+ * @relates VectorizedArray
+ */
+template <typename Number>
+inline DEAL_II_ALWAYS_INLINE
+bool
+operator == (const VectorizedArray<Number> &lhs,
```
Comment (jppelteret) [inline include/deal.II/base/vectorization.h]: FYI. This function, along with [this](https://github.com/dealii/dealii/pull/4785/files#diff-3d00cdffd465e8ebfbffae5c7a742b52R602) and [this](https://github.com/dealii/dealii/pull/4785/files#diff-3d00cdffd465e8ebfbffae5c7a742b52R613) are only necessary to ensure that these checks
```
d == internal::NumberType<Number>::value(0.0)
```
are still valid for `Number == Tensor<1,dim,VectorizedArray<othernumber>>`. I don't know how this worked before but, after explicitly setting the RHS value in the assertion, this workaround was absolutely necessary.

Comment (kronbichler) [inline include/deal.II/base/vectorization.h]: I guess we never tested `operator=(Number)` before. I guess you stumbled over a test that exposed this? In any case, it makes definitely sense to add this operator overload. :+1:

Comment (jppelteret) [inline include/deal.II/base/vectorization.h]: Actually, this caused a compilation in the main library (exposed at about 97% build completion :-/). 

Here is one such error message that this fixes / works around:
```
In file included from <>/dealii/source/matrix_free/matrix_free.cc:17:
In file included from <>/include/deal.II/matrix_free/matrix_free.templates.h:22:
In file included from <>/include/deal.II/base/tensor_product_polynomials.h:22:
<>/include/deal.II/base/tensor.h:946:13: error: invalid operands to binary expression ('const dealii::VectorizedArray<double>' and 'VectorizedArray<double>')
  Assert (d == internal::NumberType<Number>::value(0.0),
          ~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<>/include/deal.II/base/exceptions.h:330:11: note: expanded from macro 'Assert'
    if (!(cond))                                                              \
          ^~~~
<>/include/deal.II/base/tensor.h:618:10: note: in instantiation of member function 'dealii::Tensor<1, 1, dealii::VectorizedArray<double> >::operator=' requested here
      tmp=internal::NumberType<VectorizedArray<T> >::value(t);
         ^
<>/include/deal.II/base/tensor.h:631:41: note: in instantiation of member function 'dealii::internal::NumberType<dealii::Tensor<1, 1, dealii::VectorizedArray<double> > >::value' requested here
  : value(internal::NumberType<Number>::value(0.0))
                                        ^
<>/include/deal.II/base/tensor.h:376:25: note: in instantiation of member function 'dealii::Tensor<0, 1, dealii::Tensor<1, 1, dealii::VectorizedArray<double> > >::Tensor' requested here
  DEAL_II_CUDA_HOST_DEV Tensor ();
                        ^
<>/include/deal.II/matrix_free/mapping_info.templates.h:511:101: note: in instantiation of member function 'dealii::Tensor<1, 1, dealii::Tensor<1, 1, dealii::VectorizedArray<double> > >::Tensor'
      requested here
                          Tensor<1,(dim>1?dim*(dim-1)/2:1),Tensor<1,dim,VectorizedArray<Number> > > grad_upper;
                                                                                                    ^
<>/deal.II/base/aligned_vector.h:917:6: note: candidate template ignored: could not match 'AlignedVector' against 'VectorizedArray'
bool operator == (const AlignedVector<T> &lhs,
```
I'm not sure why the compiler could evaluate `Assert (d == Number(),...)` before but not `Assert (d == internal::NumberType<Number>::value(0.0),...)`.

Comment (bangerth) [inline include/deal.II/base/vectorization.h]: The error message uses alternately `VectorizedArray` and `dealii::VectorizedArray`. You may want to look into why that is so -- I bet that's the root cause for the problem.

File: include/deal.II/base/symmetric_tensor.h [hunk 7]
```diff
@@ -985,14 +996,26 @@ SymmetricTensor<rank,dim,Number>::SymmetricTensor (const Number (&array) [n_inde
...
 
 template <int rank, int dim, typename Number>
+template <typename OtherNumber>
 inline
 SymmetricTensor<rank,dim,Number> &
-SymmetricTensor<rank,dim,Number>::operator = (const Number d)
+SymmetricTensor<rank,dim,Number>::operator = (const SymmetricTensor<rank,dim,OtherNumber> &t)
 {
-  Assert (d==Number(), ExcMessage ("Only assignment with zero is allowed"));
+  data = static_cast<const base_tensor_type &>(t.data);
```
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.h]: hm, that won't work if you can't convert `Tensor<X>` into `Tensor<Y>` even if `X` can be converted to `Y`. Wouldn't you just want to do the assignment/conversion element by element?

File: include/deal.II/base/symmetric_tensor.h [hunk 8]
```diff
@@ -2096,6 +2099,42 @@ SymmetricTensor<rank,dim,Number>::serialize(Archive &ar, const unsigned int)
...
 /**
+ * Add two symmetric tensors. If possible, you should use <tt>operator +=</tt>
+ * instead since this does not need the creation of a temporary.
+ *
+ * @relates SymmetricTensor
+ */
+template <int rank, int dim, typename Number, typename OtherNumber>
+inline
+SymmetricTensor<rank, dim, typename ProductType<Number, OtherNumber>::type>
+operator + (const SymmetricTensor<rank, dim, Number>      &left,
```
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.h]: so this seems wrong -- you don't want `ProductType` but `SumType` for the return object, don't you? That *may* sometimes be the same time, but certainly not always.

File: include/deal.II/base/symmetric_tensor.h [hunk 9]
```diff
@@ -2096,6 +2099,42 @@ SymmetricTensor<rank,dim,Number>::serialize(Archive &ar, const unsigned int)
...
+/**
+ * Subtract two symmetric tensors. If possible, you should use <tt>operator -=</tt>
+ * instead since this does not need the creation of a temporary.
+ *
+ * @relates SymmetricTensor
+ */
+template <int rank, int dim, typename Number, typename OtherNumber>
+inline
+SymmetricTensor<rank, dim, typename ProductType<Number, OtherNumber>::type>
+operator - (const SymmetricTensor<rank, dim, Number>      &left,
```
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.h]: same here

File: include/deal.II/base/tensor.h [hunk 10]
```diff
@@ -585,21 +586,59 @@ class Tensor
...
 
 
 template <int dim,typename Number>
 inline
 DEAL_II_CUDA_HOST_DEV Tensor<0,dim,Number>::Tensor ()
-  : value()
+  : value(internal::NumberType<Number>::value(0.0))
 {
+  // Some auto-differentiable numbers need explicit
+  // zero initialization.
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: Move the comment above the place where you actually need it.

File: include/deal.II/base/symmetric_tensor.h (hunk 8 above)
Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: So I've been looking at this all day, and haven't had much luck at implementing a `<Sum/Addition/Subtraction>Type` that is compatible with `Sacado::Fad::DFad`. The math-ops create expressions of type `Sacado::Fad::Expr<Op>`, which are not trivially convertible from a `Sacado::Fad::DFad`, meaning that some of the initialisations implemented in the `Tensor` and `SymmetricTensor` classes do not work. However, it is possible to construct a `Sacado::Fad::DFad` directly from any `Sacado::Fad::Expr<Op>`, which is why the current implementation works (the result of any `ProductType` invoving a `Sacado::Fad::DFad` is the Sacado number).

If nobody objects, I'd like to leave this as is for now but I will try to fix it at a later stage. I agree that our current implementation is wrong, but it works fine at the moment (presumably at the expense of some performance).

Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: Maybe have a brief documentation of this discussion in the code?

File: include/deal.II/base/vectorization.h (hunk 6 above)
Comment (jppelteret) [inline include/deal.II/base/vectorization.h]: I had a brief look at the `matrix_free` headers, and nothing looks out of place to me (`#include <deal.II/base/vectorization.h>` is an indirect inclusion in `mapping_info.templates.h`). The line `mapping_info.templates.h:511` is inside `namespace internal`, and there are some direct qualifications of classes in the the `dealii` within this section of code (e.g. `dealii::Triangulation`), but it seems to have never been necessary to explicitly use `dealii::VectorizedArray` itself. 

In summary, I'm out of ideas as to what the cause of this could have been. Do you have any specific insights as to what I should be looking for, or can we consider this fixed?

File: include/deal.II/base/symmetric_tensor.h (hunk 8 above)
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.h]: Or open a github issue. I suspect that using something like `decltype(left[0][0] + right[0][0])` or similar will work. But let's not worry about it then for now if it works for you.

Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: I will open up a github issue. As I said, I __do__ want to come back to this but I don't think that its the best use of my time right now. I need more time to familiarise myself with how the Sacado AD classes and expression templates work. I'd rather get the auto-differentiation helper classes into the library first.

I'll dump the core part of my prototype code into the forthcoming issue. I did use `decltype()` etc. but as I said, having something simple like (off the top of my head)
```c++
template<typename Number1, typename Number2>
struct AdditionType
{
  typedef decltype(std::declval<Number1>() + std::declval<Number2>()) type;
};

template<typename Number1, typename Number2>
typename AdditionType<Number1,Number2>::type
add_1 (const Number1 &a, const Number2 &b)
{
   typename AdditionType<Number1,Number2>::type out;
  out = internal::NumberType<Number>::value(0.0);
   out += a;
   out += b;
   return out;
}

template<typename Number1, typename Number2>
typename AdditionType<Number1,Number2>::type
add_2 (const Number1 &a, const Number2 &b)
{
   typename AdditionType<Number1,Number2>::type out (a);
   out += b;
   return out;
}
```
does not necessarily work for all combinations of Sacado related `Number1`,`Number2` and `AdditionType::type`.

Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: Oh, and got even worst when one accumulated complex constructions of `AdditionType`s, `SumType`s and `ProductTypes`...

File: include/deal.II/base/vectorization.h (hunk 6 above)
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: I don't. If you had a test case that failed before, I could take a look. But given that you made it work (correct?) it may not be worth trying to figure out what exactly might have been going wrong.

File: include/deal.II/base/symmetric_tensor.h (hunk 8 above)
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.h]: Good plan then!

File: include/deal.II/base/vectorization.h (hunk 6 above)
Comment (jppelteret) [inline include/deal.II/base/vectorization.h]: Well, this simply prevented compilation of the library rather raised issues with a particular test. So its "self-testing" :-)

**Comment by jppelteret** [comment]
I think that I attended to everyone's comments in this PR. If its considered satisfactory, could we please merge it?

To my mind, the only possible outstanding question is whether my changes did indeed break the anisotropic test. If so, then I'll fix it up in another PR.
```
Check /home/bob/build-gcc/tests/base/anisotropic_1.debug/output /home/bob/source/tests/base/anisotropic_1.output
base/anisotropic_1.debug: DIFF failed. ------ Diff:   /home/bob/build-gcc/tests/base/anisotropic_1.debug/failing_diff
base/anisotropic_1.debug: DIFF failed. ------ First 8 lines of numdiff/diff output:
----------------
##17      #:8   <== -52.8413
##17      #:8   ==> -52.8412
@ Absolute error = 1.0000000000e-4, Relative error = 1.8924626996e-6
----------------
##18      #:8   <== -52.8413
##18      #:8   ==> -52.8412
@ Absolute error = 1.0000000000e-4, Relative error = 1.8924626996e-6
base/anisotropic_1.debug: DIFF failed. ------ First 50 lines diff output:
```

**Comment by bangerth** [comment]
That seems like the kind of problem we get if the real number is 52.84125. And yes, we can fix that up separately.

I'll take a look at the PR again later.

File: include/deal.II/base/symmetric_tensor.h [hunk 11]
```diff
@@ -915,6 +915,17 @@ namespace internal
 
 template <int rank, int dim, typename Number>
 inline
+SymmetricTensor<rank,dim,Number>::SymmetricTensor ()
+{
+  // Some auto-differentiable numbers need explicit
+  // zero initialization.
+  for (unsigned int i=0; i<base_tensor_type::dimension; ++i)
+    data[i] = internal::NumberType<Number>::value(0.0);
```
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.h]: I missed this the first time around -- for rank-4 tensors, `data` is a rank-2 tensor. Can it be initialized this way? I think you may want to try creating a `SymmetricTensor<4,dim,...some AD type...>` or maybe just using a `std::complex` type. That would probably also make for a good stand-alone testcase.

File: include/deal.II/base/symmetric_tensor.h [hunk 12]
```diff
@@ -985,14 +996,27 @@ SymmetricTensor<rank,dim,Number>::SymmetricTensor (const Number (&array) [n_inde
 
 
 template <int rank, int dim, typename Number>
+template <typename OtherNumber>
+inline
+SymmetricTensor<rank,dim,Number> &
+SymmetricTensor<rank,dim,Number>::operator = (const SymmetricTensor<rank,dim,OtherNumber> &t)
+{
+  for (unsigned int i=0; i<base_tensor_type::dimension; ++i)
+    data[i] = t.data[i];
```
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.h]: same question here about what happens for rank-4 tensors with AD types or complex types

File: include/deal.II/base/symmetric_tensor.h (hunk 11 above)
Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: Hmm... true. Good catch! I'll write some test cases as you suggested and, if there is a problem, implement an alternate strategy for initialisation.

File: include/deal.II/base/symmetric_tensor.h [hunk 13]
```diff
@@ -2096,8 +2100,57 @@ SymmetricTensor<rank,dim,Number>::serialize(Archive &ar, const unsigned int)
 
 
 /**
+ * Addition of two symmetric tensors of equal rank. The result is another
+ * SymmetricTensor that has a number type that is compatible with the
+ * operation.
+ *
+ * If possible (e.g. when @p Number and @p OtherNumber are of the same type,
+ * or if the result of <code>Number() + OtherNumber()</code> is another @p Number),
+ * you should use <tt>operator +=</tt> instead since this does not need the
```
Comment (tjhei) [inline include/deal.II/base/symmetric_tensor.h]: need -> require

File: include/deal.II/base/symmetric_tensor.h [hunk 14]
```diff
@@ -2096,8 +2100,57 @@ SymmetricTensor<rank,dim,Number>::serialize(Archive &ar, const unsigned int)
...
+
+
+/**
+ * Subtraction of two symmetric tensors of equal rank. The result is another
+ * SymmetricTensor that has a number type that is compatible with the
+ * operation.
+ *
+ * If possible (e.g. when @p Number and @p OtherNumber are of the same type,
+ * or if the result of <code>Number() + OtherNumber()</code> is another @p Number),
+ * you should use <tt>operator +=</tt> instead since this does not need the
```
Comment (tjhei) [inline include/deal.II/base/symmetric_tensor.h]: need -> require

File: include/deal.II/base/symmetric_tensor.h (hunk 11 above)
Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: This works, but we now have a test that explicitly verifies it.

File: include/deal.II/base/symmetric_tensor.h (hunk 12 above)
Comment (jppelteret) [inline include/deal.II/base/symmetric_tensor.h]: This works, but we now have a test that explicitly verifies it.

**Comment by jppelteret** [comment]
Relates #4958
